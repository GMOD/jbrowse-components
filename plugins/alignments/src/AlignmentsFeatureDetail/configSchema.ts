import { ConfigurationSchema } from '@jbrowse/core/configuration'

export const configSchema = ConfigurationSchema('AlignmentsFeatureWidget', {})
