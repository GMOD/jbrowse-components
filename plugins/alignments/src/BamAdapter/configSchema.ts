import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import { types } from 'mobx-state-tree'

export default types.late(() =>
  ConfigurationSchema(
    'BamAdapter',
    {
      bamLocation: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/my.bam' },
      },
      index: ConfigurationSchema('BamIndex', {
        indexType: {
          model: types.enumeration('IndexType', ['BAI', 'CSI']),
          type: 'stringEnum',
          defaultValue: 'BAI',
        },
        location: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.bam.bai' },
        },
      }),
      chunkSizeLimit: {
        type: 'number',
        defaultValue: 100000000,
      },
      fetchSizeLimit: {
        type: 'number',
        defaultValue: 500000000,
      },
      sequenceAdapter: {
        type: 'frozen',
        defaultValue: null,
      },
    },
    { explicitlyTyped: true },
  ),
)
