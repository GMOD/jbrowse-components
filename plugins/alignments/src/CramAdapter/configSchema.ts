import PluginManager from '@jbrowse/core/PluginManager'
import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default (pluginManager: PluginManager) => {
  const { types } = pluginManager.lib['mobx-state-tree']
  return types.late(() =>
    ConfigurationSchema(
      'CramAdapter',
      {
        cramLocation: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.cram' },
        },
        craiLocation: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.cram.crai' },
        },
        sequenceAdapter: pluginManager.pluggableConfigSchemaType('adapter'),
      },
      { explicitlyTyped: true },
    ),
  )
}
