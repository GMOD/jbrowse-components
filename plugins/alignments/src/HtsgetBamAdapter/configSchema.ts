import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types } from 'mobx-state-tree'

export default types.late(() =>
  ConfigurationSchema(
    'HtsgetBamAdapter',
    {
      htsgetBase: {
        type: 'string',
        defaultValue: '',
      },
      htsgetTrackId: {
        type: 'string',
        defaultValue: '',
      },
      sequenceAdapter: {
        type: 'frozen',
        defaultValue: null,
      },
    },
    { explicitlyTyped: true },
  ),
)
