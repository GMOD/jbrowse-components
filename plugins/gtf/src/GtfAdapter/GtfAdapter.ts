import {
  BaseFeatureDataAdapter,
  BaseOptions,
} from '@jbrowse/core/data_adapters/BaseAdapter'
import { NoAssemblyRegion } from '@jbrowse/core/util/types'
import { readConfObject } from '@jbrowse/core/configuration'
import { openLocation } from '@jbrowse/core/util/io'
import { ObservableCreate } from '@jbrowse/core/util/rxjs'
import IntervalTree from '@flatten-js/interval-tree'
import SimpleFeature, { Feature } from '@jbrowse/core/util/simpleFeature'
import { unzip } from '@gmod/bgzf-filehandle'

import gtf from '@gmod/gtf'
import { FeatureLoc, featureData } from '../util'
function isGzip(buf: Buffer) {
  return buf[0] === 31 && buf[1] === 139 && buf[2] === 8
}

export default class extends BaseFeatureDataAdapter {
  protected gtfFeatures?: Promise<{
    intervalTree: Record<string, IntervalTree>
  }>

  private async loadDataP() {
    const buffer = await openLocation(
      readConfObject(this.config, 'gtfLocation'),
      this.pluginManager,
    ).readFile()

    const buf = isGzip(buffer) ? await unzip(buffer) : buffer
    // 512MB  max chrome string length is 512MB
    if (buf.length > 536_870_888) {
      throw new Error('Data exceeds maximum string length (512MB)')
    }
    const data = new TextDecoder('utf8', { fatal: true }).decode(buf)
    const feats = gtf.parseStringSync(data, {
      parseFeatures: true,
      parseComments: false,
      parseDirectives: false,
      parseSequences: false,
    }) as FeatureLoc[][]

    const intervalTree = feats
      .flat()
      .map(
        (f, i) =>
          new SimpleFeature({
            data: featureData(f),
            id: `${this.id}-offset-${i}`,
          }),
      )
      .reduce((acc, obj) => {
        const key = obj.get('refName')
        if (!acc[key]) {
          acc[key] = new IntervalTree()
        }
        acc[key].insert([obj.get('start'), obj.get('end')], obj)
        return acc
      }, {} as Record<string, IntervalTree>)
    return { intervalTree }
  }

  private async loadData() {
    if (!this.gtfFeatures) {
      this.gtfFeatures = this.loadDataP().catch(e => {
        this.gtfFeatures = undefined
        throw e
      })
    }

    return this.gtfFeatures
  }

  public async getRefNames(opts: BaseOptions = {}) {
    const { intervalTree } = await this.loadData()
    return Object.keys(intervalTree)
  }

  public getFeatures(query: NoAssemblyRegion, opts: BaseOptions = {}) {
    return ObservableCreate<Feature>(async observer => {
      try {
        const { start, end, refName } = query
        const { intervalTree } = await this.loadData()
        intervalTree[refName]
          ?.search([start, end])
          .forEach(f => observer.next(f))
        observer.complete()
      } catch (e) {
        observer.error(e)
      }
    }, opts.signal)
  }
  public freeResources(/* { region } */) {}
}
