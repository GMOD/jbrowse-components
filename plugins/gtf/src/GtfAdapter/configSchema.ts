import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'GtfAdapter',
  {
    gtfLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.gtf', locationType: 'UriLocation' },
    },
  },
  { explicitlyTyped: true },
)
