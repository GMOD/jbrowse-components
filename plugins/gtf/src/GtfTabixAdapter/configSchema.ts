import { types } from 'mobx-state-tree'
import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'GtfTabixAdapter',
  {
    gtfGzLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.gtf.gz' },
    },
    index: ConfigurationSchema('GtfTabixIndex', {
      indexType: {
        model: types.enumeration('IndexType', ['TBI', 'CSI']),
        type: 'stringEnum',
        defaultValue: 'TBI',
      },
      location: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/my.gtf.gz.tbi' },
      },
    }),
    dontRedispatch: {
      type: 'stringArray',
      defaultValue: ['chromosome', 'region'],
    },
  },
  { explicitlyTyped: true },
)
