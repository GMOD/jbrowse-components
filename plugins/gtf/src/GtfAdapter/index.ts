import { AdapterType } from '@jbrowse/core/pluggableElementTypes'

import configSchema from './configSchema'

import type PluginManager from '@jbrowse/core/PluginManager'

export default function GtfAdapterF(pluginManager: PluginManager) {
  pluginManager.addAdapterType(
    () =>
      new AdapterType({
        name: 'GtfAdapter',
        displayName: 'GTF adapter',
        configSchema,
        getAdapterClass: () => import('./GtfAdapter').then(r => r.default),
      }),
  )
}
