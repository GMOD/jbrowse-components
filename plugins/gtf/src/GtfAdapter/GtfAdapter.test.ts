import { toArray } from 'rxjs/operators'
import configSchema from './configSchema'
import GtfAdapter from './GtfAdapter'

import { TextDecoder } from 'web-encoding'
window.TextDecoder = TextDecoder

describe('adapter can fetch features from volvox.sorted.gtf', () => {
  let adapter: GtfAdapter
  beforeEach(() => {
    adapter = new GtfAdapter(
      configSchema.create({
        gtfLocation: {
          localPath: require.resolve('../test_data/volvox.sorted.gtf'),
        },
      }),
    )
  })
  it('test getfeatures on gtf plain text adapter', async () => {
    const features = adapter.getFeatures({
      refName: 'ctgA',
      start: 0,
      end: 100000,
    })
    expect(await adapter.hasDataForRefName('ctgA')).toBe(true)
    expect(await adapter.hasDataForRefName('ctgB')).toBe(false)
    const featuresArray = await features.pipe(toArray()).toPromise()
    // There are only 4 features in ctgB
    expect(featuresArray.length).toBe(6)
    const featuresJsonArray = featuresArray.map(f => f.toJSON())
    expect(featuresJsonArray).toMatchSnapshot()
  })
})

test('can instantiate new GtfAdapter and check for demo data', async () => {
  const demoAdapter = new GtfAdapter(
    configSchema.create({
      gtfLocation: {
        localPath: require.resolve('../test_data/demo.gtf'),
      },
    }),
  )
  expect(await demoAdapter.hasDataForRefName('GeneScaffold_1')).toBe(true)
  expect(await demoAdapter.hasDataForRefName('GeneScaffold_10')).toBe(true)
  expect(await demoAdapter.hasDataForRefName('GeneScaffold_11')).toBe(false)

  const features = demoAdapter.getFeatures({
    refName: 'GeneScaffold_1',
    start: 0,
    end: 1100000,
  })
  const featuresArray = await features.pipe(toArray()).toPromise()
  const featuresJsonArray = featuresArray.map(f => f.toJSON())
  // ENSVPAT00000000407
  expect(featuresArray.length).toEqual(1)
  expect(featuresJsonArray).toMatchSnapshot()
})
