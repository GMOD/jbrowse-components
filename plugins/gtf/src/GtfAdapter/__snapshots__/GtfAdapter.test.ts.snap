// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from volvox.sorted.gtf test getfeatures on gtf plain text adapter 1`] = `
Array [
  Object {
    "end": 9000,
    "gene_id": "EDEN",
    "gene_name": "EDEN",
    "name": "EDEN.1",
    "phase": 0,
    "refName": "ctgA",
    "source": "example",
    "start": 1049,
    "strand": 1,
    "subfeatures": Array [
      Object {
        "end": 1500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 1049,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "exon",
        "uniqueId": "test-offset-0-0",
      },
      Object {
        "end": 1200,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 1049,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "five_prime_UTR",
        "uniqueId": "test-offset-0-1",
      },
      Object {
        "end": 1500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 1200,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "CDS",
        "uniqueId": "test-offset-0-2",
      },
      Object {
        "end": 3902,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 2999,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "CDS",
        "uniqueId": "test-offset-0-3",
      },
      Object {
        "end": 3902,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 2999,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "exon",
        "uniqueId": "test-offset-0-4",
      },
      Object {
        "end": 5500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 4999,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "CDS",
        "uniqueId": "test-offset-0-5",
      },
      Object {
        "end": 5500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 4999,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "exon",
        "uniqueId": "test-offset-0-6",
      },
      Object {
        "end": 7608,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 6999,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "CDS",
        "uniqueId": "test-offset-0-7",
      },
      Object {
        "end": 9000,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 6999,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "exon",
        "uniqueId": "test-offset-0-8",
      },
      Object {
        "end": 9000,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.1",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 7608,
        "strand": 1,
        "transcript_id": "EDEN.1",
        "type": "three_prime_UTR",
        "uniqueId": "test-offset-0-9",
      },
    ],
    "transcript_id": "EDEN.1",
    "type": "transcript",
    "uniqueId": "test-offset-0",
  },
  Object {
    "end": 9000,
    "gene_id": "EDEN",
    "gene_name": "EDEN",
    "name": "EDEN.2",
    "phase": 0,
    "refName": "ctgA",
    "source": "example",
    "start": 1049,
    "strand": 1,
    "subfeatures": Array [
      Object {
        "end": 1500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 1049,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "exon",
        "uniqueId": "test-offset-1-0",
      },
      Object {
        "end": 1200,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 1049,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "five_prime_UTR",
        "uniqueId": "test-offset-1-1",
      },
      Object {
        "end": 1500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 1200,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "CDS",
        "uniqueId": "test-offset-1-2",
      },
      Object {
        "end": 5500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 4999,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "CDS",
        "uniqueId": "test-offset-1-3",
      },
      Object {
        "end": 5500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 4999,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "exon",
        "uniqueId": "test-offset-1-4",
      },
      Object {
        "end": 7608,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 6999,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "CDS",
        "uniqueId": "test-offset-1-5",
      },
      Object {
        "end": 9000,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 6999,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "exon",
        "uniqueId": "test-offset-1-6",
      },
      Object {
        "end": 9000,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.2",
        "parentId": "test-offset-1",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 7608,
        "strand": 1,
        "transcript_id": "EDEN.2",
        "type": "three_prime_UTR",
        "uniqueId": "test-offset-1-7",
      },
    ],
    "transcript_id": "EDEN.2",
    "type": "transcript",
    "uniqueId": "test-offset-1",
  },
  Object {
    "end": 9000,
    "gene_id": "EDEN",
    "gene_name": "EDEN",
    "name": "EDEN.3",
    "phase": 0,
    "refName": "ctgA",
    "source": "example",
    "start": 1299,
    "strand": 1,
    "subfeatures": Array [
      Object {
        "end": 1500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 1299,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "exon",
        "uniqueId": "test-offset-2-0",
      },
      Object {
        "end": 1500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 1,
        "refName": "ctgA",
        "source": "example",
        "start": 1299,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "five_prime_UTR",
        "uniqueId": "test-offset-2-1",
      },
      Object {
        "end": 3902,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 2999,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "exon",
        "uniqueId": "test-offset-2-2",
      },
      Object {
        "end": 3300,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 1,
        "refName": "ctgA",
        "source": "example",
        "start": 2999,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "five_prime_UTR",
        "uniqueId": "test-offset-2-3",
      },
      Object {
        "end": 3902,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 3300,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "CDS",
        "uniqueId": "test-offset-2-4",
      },
      Object {
        "end": 5500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 1,
        "refName": "ctgA",
        "source": "example",
        "start": 4999,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "CDS",
        "uniqueId": "test-offset-2-5",
      },
      Object {
        "end": 5500,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 4999,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "exon",
        "uniqueId": "test-offset-2-6",
      },
      Object {
        "end": 7600,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 1,
        "refName": "ctgA",
        "source": "example",
        "start": 6999,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "CDS",
        "uniqueId": "test-offset-2-7",
      },
      Object {
        "end": 9000,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 0,
        "refName": "ctgA",
        "source": "example",
        "start": 6999,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "exon",
        "uniqueId": "test-offset-2-8",
      },
      Object {
        "end": 9000,
        "gene_id": "EDEN",
        "gene_name": "EDEN",
        "name": "EDEN.3",
        "parentId": "test-offset-2",
        "phase": 1,
        "refName": "ctgA",
        "source": "example",
        "start": 7600,
        "strand": 1,
        "transcript_id": "EDEN.3",
        "type": "three_prime_UTR",
        "uniqueId": "test-offset-2-9",
      },
    ],
    "transcript_id": "EDEN.3",
    "type": "transcript",
    "uniqueId": "test-offset-2",
  },
  Object {
    "end": 11500,
    "name": "Apple1",
    "phase": 0,
    "refName": "ctgA",
    "source": "bare_predicted",
    "start": 9999,
    "strand": 1,
    "subfeatures": Array [
      Object {
        "end": 11500,
        "name": "Apple1",
        "parentId": "test-offset-3",
        "phase": 0,
        "refName": "ctgA",
        "source": "bare_predicted",
        "start": 9999,
        "strand": 1,
        "transcript_id": "Apple1",
        "type": "CDS",
        "uniqueId": "test-offset-3-0",
      },
    ],
    "transcript_id": "Apple1",
    "type": "transcript",
    "uniqueId": "test-offset-3",
  },
  Object {
    "end": 17000,
    "name": "cds-Apple2",
    "phase": 0,
    "refName": "ctgA",
    "source": "predicted",
    "start": 13499,
    "strand": 1,
    "subfeatures": Array [
      Object {
        "end": 13800,
        "name": "cds-Apple2",
        "parentId": "test-offset-4",
        "phase": 0,
        "refName": "ctgA",
        "source": "predicted",
        "start": 13499,
        "strand": 1,
        "transcript_id": "cds-Apple2",
        "type": "CDS",
        "uniqueId": "test-offset-4-0",
      },
      Object {
        "end": 15500,
        "name": "cds-Apple2",
        "parentId": "test-offset-4",
        "phase": 1,
        "refName": "ctgA",
        "source": "predicted",
        "start": 14999,
        "strand": 1,
        "transcript_id": "cds-Apple2",
        "type": "CDS",
        "uniqueId": "test-offset-4-1",
      },
      Object {
        "end": 17000,
        "name": "cds-Apple2",
        "parentId": "test-offset-4",
        "phase": 2,
        "refName": "ctgA",
        "source": "predicted",
        "start": 16499,
        "strand": 1,
        "transcript_id": "cds-Apple2",
        "type": "CDS",
        "uniqueId": "test-offset-4-2",
      },
    ],
    "transcript_id": "cds-Apple2",
    "type": "transcript",
    "uniqueId": "test-offset-4",
  },
  Object {
    "end": 23000,
    "name": "rna-Apple3",
    "phase": 0,
    "refName": "ctgA",
    "source": "exonerate",
    "start": 17399,
    "strand": 1,
    "subfeatures": Array [
      Object {
        "end": 17999,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 17399,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "UTR",
        "uniqueId": "test-offset-5-0",
      },
      Object {
        "end": 18800,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 17399,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "exon",
        "uniqueId": "test-offset-5-1",
      },
      Object {
        "end": 18800,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 17999,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "CDS",
        "uniqueId": "test-offset-5-2",
      },
      Object {
        "end": 19500,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 18999,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "CDS",
        "uniqueId": "test-offset-5-3",
      },
      Object {
        "end": 19500,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 18999,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "exon",
        "uniqueId": "test-offset-5-4",
      },
      Object {
        "end": 21200,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 20999,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "CDS",
        "uniqueId": "test-offset-5-5",
      },
      Object {
        "end": 23000,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 20999,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "exon",
        "uniqueId": "test-offset-5-6",
      },
      Object {
        "end": 23000,
        "name": "rna-Apple3",
        "parentId": "test-offset-5",
        "phase": 0,
        "refName": "ctgA",
        "source": "exonerate",
        "start": 21200,
        "strand": 1,
        "transcript_id": "rna-Apple3",
        "type": "UTR",
        "uniqueId": "test-offset-5-7",
      },
    ],
    "transcript_id": "rna-Apple3",
    "type": "transcript",
    "uniqueId": "test-offset-5",
  },
]
`;

exports[`can instantiate new GtfAdapter and check for demo data 1`] = `
Array [
  Object {
    "end": 107987,
    "gene_biotype": "pseudogene",
    "gene_id": "ENSVPAG00000000407",
    "gene_source": "ensembl",
    "gene_version": "1",
    "name": "ENSVPAT00000000407",
    "phase": 0,
    "refName": "GeneScaffold_1",
    "source": "ensembl",
    "start": 107688,
    "strand": 1,
    "subfeatures": Array [
      Object {
        "end": 107715,
        "exon_id": "ENSVPAE00000004844",
        "exon_number": "1",
        "exon_version": "1",
        "gene_biotype": "pseudogene",
        "gene_id": "ENSVPAG00000000407",
        "gene_source": "ensembl",
        "gene_version": "1",
        "name": "ENSVPAT00000000407",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "GeneScaffold_1",
        "source": "ensembl",
        "start": 107688,
        "strand": 1,
        "transcript_biotype": "pseudogene",
        "transcript_id": "ENSVPAT00000000407",
        "transcript_source": "ensembl",
        "transcript_version": "1",
        "type": "exon",
        "uniqueId": "test-offset-0-0",
      },
      Object {
        "end": 107812,
        "exon_id": "ENSVPAE00000004847",
        "exon_number": "2",
        "exon_version": "1",
        "gene_biotype": "pseudogene",
        "gene_id": "ENSVPAG00000000407",
        "gene_source": "ensembl",
        "gene_version": "1",
        "name": "ENSVPAT00000000407",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "GeneScaffold_1",
        "source": "ensembl",
        "start": 107716,
        "strand": 1,
        "transcript_biotype": "pseudogene",
        "transcript_id": "ENSVPAT00000000407",
        "transcript_source": "ensembl",
        "transcript_version": "1",
        "type": "exon",
        "uniqueId": "test-offset-0-1",
      },
      Object {
        "end": 107877,
        "exon_id": "ENSVPAE00000004849",
        "exon_number": "3",
        "exon_version": "1",
        "gene_biotype": "pseudogene",
        "gene_id": "ENSVPAG00000000407",
        "gene_source": "ensembl",
        "gene_version": "1",
        "name": "ENSVPAT00000000407",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "GeneScaffold_1",
        "source": "ensembl",
        "start": 107823,
        "strand": 1,
        "transcript_biotype": "pseudogene",
        "transcript_id": "ENSVPAT00000000407",
        "transcript_source": "ensembl",
        "transcript_version": "1",
        "type": "exon",
        "uniqueId": "test-offset-0-2",
      },
      Object {
        "end": 107895,
        "exon_id": "ENSVPAE00000156282",
        "exon_number": "4",
        "exon_version": "1",
        "gene_biotype": "pseudogene",
        "gene_id": "ENSVPAG00000000407",
        "gene_source": "ensembl",
        "gene_version": "1",
        "name": "ENSVPAT00000000407",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "GeneScaffold_1",
        "source": "ensembl",
        "start": 107880,
        "strand": 1,
        "transcript_biotype": "pseudogene",
        "transcript_id": "ENSVPAT00000000407",
        "transcript_source": "ensembl",
        "transcript_version": "1",
        "type": "exon",
        "uniqueId": "test-offset-0-3",
      },
      Object {
        "end": 107917,
        "exon_id": "ENSVPAE00000188321",
        "exon_number": "5",
        "exon_version": "1",
        "gene_biotype": "pseudogene",
        "gene_id": "ENSVPAG00000000407",
        "gene_source": "ensembl",
        "gene_version": "1",
        "name": "ENSVPAT00000000407",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "GeneScaffold_1",
        "source": "ensembl",
        "start": 107899,
        "strand": 1,
        "transcript_biotype": "pseudogene",
        "transcript_id": "ENSVPAT00000000407",
        "transcript_source": "ensembl",
        "transcript_version": "1",
        "type": "exon",
        "uniqueId": "test-offset-0-4",
      },
      Object {
        "end": 107987,
        "exon_id": "ENSVPAE00000004854",
        "exon_number": "6",
        "exon_version": "1",
        "gene_biotype": "pseudogene",
        "gene_id": "ENSVPAG00000000407",
        "gene_source": "ensembl",
        "gene_version": "1",
        "name": "ENSVPAT00000000407",
        "parentId": "test-offset-0",
        "phase": 0,
        "refName": "GeneScaffold_1",
        "source": "ensembl",
        "start": 107918,
        "strand": 1,
        "transcript_biotype": "pseudogene",
        "transcript_id": "ENSVPAT00000000407",
        "transcript_source": "ensembl",
        "transcript_version": "1",
        "type": "exon",
        "uniqueId": "test-offset-0-5",
      },
    ],
    "transcript_biotype": "pseudogene",
    "transcript_id": "ENSVPAT00000000407",
    "transcript_source": "ensembl",
    "transcript_version": "1",
    "type": "transcript",
    "uniqueId": "test-offset-0",
  },
]
`;
